File: src/DiagonalTraverse/DiagonalTraverse.java
class Solution {
    int m;
    int n;
    public int[] findDiagonalOrder(int[][] mat) {
        m = mat.length;
        n = mat[0].length;
        int[] res = new int[m * n];
        
        //Traverse the list
        int row = 0, col = 0, index = 0;
        
        while(row < m && col < n){
            res[index++] = mat[row][col];
            if(traverseUp(row, col)){
                if(withBoundary(row - 1, col + 1)){
                    row--;
                    col++;
                }else if(withBoundary(row, col + 1)){
                    col++;
                }else{
                    row++;
                }
            }else{
                if(withBoundary(row + 1, col - 1)){
                    row++;
                    col--;
                }else if(withBoundary(row + 1, col)){
                    row++;
                }else{
                    col++;
                }
            }
        }
        return res;
    }
    private boolean traverseUp(int row, int col){
        return (row + col) % 2 == 0;
    }
    private boolean withBoundary(int row, int col){
        if(row < 0 || m <= row) return false;
        if(col < 0 || n <= col) return false;
        return true;
    }
}

File: src/Array/DiagonalTraverse/DiagonalTraverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:

        m = len(mat)
        n = len(mat[0])

        def is_traverse_up(row, col):
            # if even number we are traversing up right
            return (row + col) % 2 == 0

        def is_with_in_boundary(row, col):
            return 0 <= row and 0 <= col and row < m and col < n

        res = []
        row = col = 0

        for i in range(m * n):
            res.append(mat[row][col])

            if is_traverse_up(row, col):
                if is_with_in_boundary(row - 1, col + 1):
                    row -= 1
                    col += 1
                elif is_with_in_boundary(row, col + 1):
                    col += 1
                else:
                    row += 1
            else:
                if is_with_in_boundary(row+1, col-1):
                    row += 1
                    col -= 1
                elif is_with_in_boundary(row+1, col):
                    row += 1
                else:
                    col += 1
        return res
